### PR TITLE
reference sheets

### PR DESCRIPTION
**Pull request by kishan243**
pr for sensors, file structure, and command based

### CODE REVIEW

File: reference-sheets/Sensors.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+### [Absolute Encoders](https://docs.wpilib.org/en/stable/docs/software/hardware-apis/sensors/encoders-software.html#encoders-software)
+
+Has a set zero point. Will always know where it is, even between code deploys.
+
+## [Gyroscope](https://docs.wpilib.org/en/stable/docs/software/hardware-apis/sensors/gyros-software.html)
+
+It measures the rate of rotation of whatever plane it is on (and sometimes its relative axes). Usually found on the base of the drivetrain.
+
+## [Beambreaks]()
```
Comment (Yxhej) [inline reference-sheets/Sensors.md:23]: empty link, honestly not sure what to link to here; maybe nothing at all

File: reference-sheets/GitGuide.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+Review and discuss changes before merging them into the main branch. This helps catch errors, improve code quality, and ensure everyone is aligned.
+
+## Cheat Sheet
+
+Simple list of all the most commonly used commands.
+
+#### Cloning
+`git clone <repository-link | repository-name> [repository-folder]`
+
+Copies the repo (repository) from github and puts it in a folder on your computer. It will put it in a folder with the name of the repository unless you specify a folder afterwards.
```
Comment (Yxhej) [inline reference-sheets/GitGuide.md:28]: Note that the default place of creation (no folder location arg) is the terminal's working directory.

File: reference-sheets/GitGuide.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+
+#### Committing
+`git commit [-m <message>]`
+
+Every commit in GitHub you can think of as saving your progress. Although we’d like for our commit messages to be as descriptive as possible, being too detailed would be inappropriate as nobody would want to read such long commit messages. Try to limit your message lengths to at most a sentence, or make it like a title.
+
+`git commit -m “decreased code duplication in Robot.java”`
+
+However, before you can commit, you need to “mark” the files you would like to save. Only marked files will be saved. To mark a file you will need to  git add  modified files. For example, to add all changed files in your current directory, you may use:
```
Comment (Yxhej) [inline reference-sheets/GitGuide.md:38]:
```suggestion
However, before you can commit, you need to “mark” the files you would like to save. Only marked files will be saved. To mark a file you will need to `git add` modified files. For example, to add all changed files in your current directory, you may use: 
```

File: reference-sheets/GitGuide.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+`git commit -m “decreased code duplication in Robot.java”`
+
+However, before you can commit, you need to “mark” the files you would like to save. Only marked files will be saved. To mark a file you will need to  git add  modified files. For example, to add all changed files in your current directory, you may use: 
+
+`git add .`
+
+#### Pulling
+`git pull`
+
+Copies any new commits that are on the remote GitHub repo to your local repo.
```
Comment (Yxhej) [inline reference-sheets/GitGuide.md:45]:
```suggestion
Copies any new commits that are on the remote GitHub repo to your local repo and attempts to merge its changes into your local repository.
```

File: reference-sheets/GitGuide.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+Moves you to the specified branch.
+
+`git checkout -b <new-branch-name>`
+
+Creates that branch and moves you to it. This is equivalent to:
+
+- `git branch ball-counter`
+- `git checkout ball-counter`
+
+#### Merging
```
Comment (Yxhej) [inline reference-sheets/GitGuide.md:78]: Talk about merge conflicts... I reckon it'll be a problem for people working on different machines at home

File: reference-sheets/Sensors.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+Detect if something has passed through them. Two key components: a part that shoots a ray of light, and a receiver. When the receiver no longer detects light, a signal is returned.
+
+For our program, beambreak sensors return true when unblocked, and false when blocked. **THIS IS SUBJECT TO CHANGE.**
```
Comment (Yxhej) [inline reference-sheets/Sensors.md:27]:
```suggestion
For our specific hardware, beambreak sensors return true when unblocked, and false when blocked.
```

File: reference-sheets/FileStructure.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+- `Robot.java`: this is the file which the entire robot code really centers upon. This file will contain instances of all of the subsystems. It contains the Xbox controllers. It runs the CommandScheduler. In short, `Robot.java` is where everything comes together. In fact, all that the `Main.java` file really does is start up `Robot.java`.
+
+As you go through a project, you will create more directories and files within `main/java/robot` for your robot code.
+
+Next up, we'll go over `test/java/robot`. This directory only has one file currently:
+- `RobotTest.java`: If you look at the contents of this file, it is a test class with a single test called `initialize`. All that this test does is create a new instance of `Robot.java`. So this test will only fail if initializing the robot throws an error. This tiny little test is actually very important, because it will catch if there is a `NullPointerException` in our code (if we try to use something that isn't initialized).
+
+Throughout a project, You will add files to `test/java/robot` to test the new classes that you create in `main/java/robot`. And you'll be testing more specific behavior than `RobotTest.java` does!
```
Comment (Yxhej) [inline reference-sheets/FileStructure.md:36]:
```suggestion
In a project, you will add files to `test/java/robot` to test the new classes that you create in `main/java/robot`. And you'll be testing more specific behavior than `RobotTest.java` does!
```

File: reference-sheets/FileStructure.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+Now, let's look at a tree that includes the actual files as well as directories:
+
+![](/images/sciguides-robot-base-file-tree.png)
+
+There's a lot here, but we'll go through it piece by piece.
+
+One thing you might notice is that the `lib` folders are very populated. That's because one of the main points of using this template is that it has library code from the SciBorgs, which you can use in your project. You will not be writing code in the `lib` folders, but you will be using the utilities from those folders. But you don't have to worry about that right now, we'll introduce specific library code when it is relevant.
+
+So we're just going to focus on the contents of the `robot` directories. Let's start with `main/java/robot`:
```
Comment (Yxhej) [inline reference-sheets/FileStructure.md:22]: Section is a bit wordy, "this is a file" can be removed and is obvious but its pedantic but should be quick enough

File: reference-sheets/CommandBased.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+Commands represent robot actions with the hardware. They should be used for actions that may interfere with how the robot is run (i.e actual robot movement or changing PID constants.)
+
+Generally when we create commands, we do so using preexisting types of commands, and a set of helpful static methods in a class called `Commands` that return Commands. (Theoretically, you could also make a whole class for each command, but that's almost never a good idea).
+
+Before we talk about types of commands, let's quickly go over what the technical definition of a Command is. The class `Command` is what all Command classes inherit from, and it has four primary methods that different Command classes override in order to define their behavior:
+- `public void initialize()`
+	- Called when the command is started
+- `public void execute()`
+	- Called every tic (every 0.02 seconds) while the command is running
```
Comment (Yxhej) [inline reference-sheets/CommandBased.md:31]:
```suggestion
	- Called every tick (every 0.02 seconds) while the command is running
```

File: reference-sheets/CommandBased.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+- `public void end(boolean interrupted)`
+	- Called when the command is ended
+	- Commands can end either because their end condition is met or because they are interrupted by another command on the same subsystem. `end` takes whether or not the command has been interrupted as an input, so that you can change the end behavior of a command based on whether it reached its end condition.
+- `public boolean isFinished()`
+	- This is the end condition for a command. It is called each tic after a command has been executed, and if it is `isFinished` returns `true`, the command is un-scheduled and `end(false)` is called (`false` because the command has not been interrupted).
```
Comment (Yxhej) [inline reference-sheets/CommandBased.md:36]:
```suggestion
	- This is the end condition for a command. It is called each tick after a command has been executed, and if it is `isFinished` returns `true`, the command is un-scheduled and `end(false)` is called (`false` because the command has not been interrupted).
```

Comment (Yxhej) [reply reference-sheets/CommandBased.md:36]: tick , line 43,

**Review by Yxhej — CHANGES_REQUESTED** [review]

File: reference-sheets/Simulation.md [hunk 11]
```diff
@@ -0,0 +1,34 @@
+# Robot Simulation
+
+Simulation in WPILib allows for code and logic to be tested onboard your computer, rather than physical hardware. It's exceptionally useful for when construction and electronics have not finished the robot, but you have code to run!
```
Comment (sigalrmp) [inline reference-sheets/Simulation.md:3]:
```suggestion
Simulation in WPILib allows for code and logic to be tested onboard your computer, rather than with physical hardware. It's exceptionally useful for when mechanical hasn't finished the robot, but you have code to run!
```

File: reference-sheets/Simulation.md [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+- [The Simulation GUI](https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/robot-simulation)
+- [WPILib's simulation classes](#simulation-classes)
+  - [Physics simulators](https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/robot-simulation/physics-sim.html)
+  - [Mechanism2d](https://docs.wpilib.org/en/stable/docs/software/dashboards/glass/mech2d-widget.html)
+  - [Field2d](https://docs.wpilib.org/en/stable/docs/software/dashboards/glass/field2d-widget.html)
+- [Logging & Dashboards](#logging--dashboards)
+
+## Logging & Dashboards
+
+In testing, it is common to want to directly observe robot measurements and values in order to tune your systems or debug. This is also incredibly important when working with simulation, as you otherwise have no reference to what is going on without a physical robot.
```
Comment (sigalrmp) [inline reference-sheets/Simulation.md:16]: talk about viewing past logs as well -- that's as important as looking at data while you're testing

File: reference-sheets/Telemetry.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+Only classes that implement `Sendable` can be sent over NetworkTables; see [here](https://docs.wpilib.org/en/stable/docs/software/telemetry/robot-telemetry-with-sendable.html#what-is-sendable) for more details.
+
+For more in-depth information on the inner workings, visit [the docs](https://docs.wpilib.org/en/stable/docs/software/telemetry/telemetry.html).
+
+## Logging libraries
+
+Logging libraries, third-party or WPILib-made, are not monolithic. Some are annotation-based (using Java `@Annotations`), while others are framework-based. All are unique; a nice list of them can be found [here](https://docs.wpilib.org/en/stable/docs/software/telemetry/3rd-party-libraries.html).
+
+- [Monologue](https://github.com/shueja/Monologue/wiki) (we use this!)
```
Comment (sigalrmp) [inline reference-sheets/Telemetry.md:17]: worth saying what we use

File: reference-sheets/Telemetry.md [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+- [Epilogue](https://docs.wpilib.org/pt/latest/docs/software/telemetry/robot-telemetry-with-annotations.html) (official WPILib)
+- [AdvantageKit](https://github.com/Mechanical-Advantage/AdvantageKit/blob/main/docs/WHAT-IS-ADVANTAGEKIT.md)
+- [URCL](https://github.com/Mechanical-Advantage/URCL)
+
+## Dashboards
+
+There are two types of dashboards: driving and programming. We'll be talking about the ones related to debugging.
```
Comment (sigalrmp) [inline reference-sheets/Telemetry.md:24]: i think it's not totally clear from this that debugging and programming are the same

**Review by sigalrmp — CHANGES_REQUESTED** [review]

**Review by Yxhej — APPROVED** [review]

**Review by sigalrmp — APPROVED** [review]
